### DISCUSSION TITLE
about Applicability

### DISCUSSION THREAD

**Discussion by Andres-Nju** [rust-lang/rust-clippy#9994] [Q&A]
I got 2 questions:

I saw the applicability of some lints is "unresolved", and I clicked the weblink provided but couldn't find such variant in the official doc.
There are only these:
pub enum Applicability {
MachineApplicable,
MaybeIncorrect,
HasPlaceholders,
Unspecified,
}
So I wonder what does this "Applicability" mean? or if it is a little bit different from those of the lints which are provided by rustc?

I also wonder whether the applicability is related to the capability of fixing codes automatically with the lint's suggestions?
For example, applicability "unresolved" means this kind of smell/bad code cannot be fixed automatically by using "cargo clippy --fix", or it just means the process of fixing can still be done, but the substitution codes still have problems?

**Comment by xFrednet** [comment]
The applicability, on Clippy's lint list, gets extracted from the source code. Occasionally, it's unable to determine the Applicability, in that case "unresolved" will be displayed.

> 2. I also wonder whether the applicability is related to the capability of fixing codes automatically with the lint's suggestions?

Yes, this is exactly what the applicability says.

**Reply by flip1995** [reply]
If the applicability is `Unresolved` it doesn't necessarily mean that the lint won't be auto fixable. It could also mean that in some cases the suggestion could be applied automatically, but in others it can't.

- `MachineApplicable` means that the suggestion is safe to automatically apply. This are all suggestions that are applied with `cargo clippy --fix`
- `MaybeIncorrect` means that when applying the suggestion it might (doesn't have to) lead to new errors (for different reasons). Those are only applied, if you specify some hidden env var called something with `YOLO`.  For a bit more information about this and how it might be handled in the future, see: https://github.com/rust-lang/rustfix/issues/200
- `HasPlaceholders` are suggestions that (most) definitely will lead to errors, because it will insert at least one placeholder, for example for a variable name.
- `Unspecified` means that no one thought about the actual applicability of the lint yet. Those can't be applied by `cargo clippy --fix`.
